include/mbedtls/pk_info.h

```diff
@@ -0,0 +1,264 @@
+ * that can be used for all keys of a given type should use alternative
+ * ("xxx_alt") interfaces instead. This interface allows using different
+ * engines for each key.
+ *
+ * An engine for asymmetric cryptography must implement the interface
+ * described in this structure. The interface for the engine may be
+ * exposed in one of two ways:
+ *
+ * - Declare the mbedtls_pk_info_t structure and instruct users to call
+ *   mbedtls_pk_setup with that structure.
+ * - Keep the mbedtls_pk_info_t structure hidden and declare a function
+ *   to call instead of mbedtls_pk_setup. This function should have an
+ *   interface of the form
+ *    `int mbedtls_pk_setup_myengine(mbedtls_pk_context *, ...)`
+ *   where the extra parameters depend on the engine, e.g. handles to keys
+ *   stored in an external cryptographic module.
+ *
+ * Unless otherwise indicated, functions returning int must return an
+ * Mbed TLS status code, either 0 for success or a negative value to indicate
+ * an error. It is recommended to use the MBEDTLS_ERR_PK_XXX error codes
+ * defined in pk.h.
+ *
+ * Some methods are optional; this is clearly indicated in their description.
+ * If a method is optional, then an opaque key implementation may put NULL
+ * in the corresponding field. The corresponding function in pk.h will
+ * return MBEDTLS_ERR_PK_TYPE_MISMATCH in this case.
+ *
+ * \note If you are using the PK interface to perform operations on
+ * keys, call the functions in pk.h. The interface in this file should only
+ * be used by implementers of opaque key engines.
+ *
+ * \warning: Do not declare this structure directly! It may be extended in
+ * future* versions of Mbed TLS. Call the macro
+ * MBEDTLS_PK_OPAQUE_INFO_1() or MBEDTLS_PK_OPAQUE_INFO_ASYNC_1() instead.
+ * These macros are guaranteed to take parameters with the same type
+ * and semantics as previous versions and fill any new field of the
+ * structure with sensible values.
+ */
+struct mbedtls_pk_info_t
+{
+    /** Key pair type.
+     *
+     * mbedtls_pk_get_type() returns this value.
+     *
+     * For transparent keys, this contains an indication of supported
+     * algorithms. For opaque keys, this is \c MBEDTLS_PK_OPAQUE. */
+    mbedtls_pk_type_t type;
+
+    /** Type name.
+     *
+     * mbedtls_pk_get_name() returns this value. It must be a
+     * null-terminated string.
+     *
+     * For transparent keys, this reflects the key type. For opaque keys,
+     * this reflects the cryptographic module driver. */
+    const char *name;
+
+    /** Get key size in bits.
+     *
+     * mbedtls_pk_get_bitlen() returns this value.
+     *
+     * This function cannot fail. */
+    size_t (*get_bitlen)( const void *ctx );
+
+    /** Tell if the context implements this type (e.g.\ ECKEY can do ECDSA).
+     *
+     * mbedtls_pk_can_do() calls this function.
+     *
+     * This function is only based on the key type. It does not take any
+     * usage restrictions into account. */
```

Comment:
I am not fully comfortable with this description, as currently `MBEDTLS_PK_ECDSA` can do only `MBEDTLS_PK_ECDSA`, so the usage restriction to signatures is being considered.

Reply:
@gilles-peskine-arm @AndrzejKurek Ping.

Reply:
Hmmm. This documentation is mostly intended for people who write opaque engines. The case of ECDH vs ECDSA is internal to the library and not implemented consistently.

It would be useful to mention this special case for the sake of library maintainers, but I don't want to confuse people who write opaque engines. How about mentioning it in a non-Doxygen comment?

Reply:
What is the reason you want to exclude usage restrictions here?

Reply:
It's impossible to express usage restrictions since there's no way to encode the precise usage in the parameters. There's only the key type, and a dubious distinction between “ECKEY”, “ECDH” and “ECDSA”. This doesn't even distinguish between ECDSA signature and verification, so a public ECC key “can do” ECDSA.

Is the sense of your question, should opaque engines reply “yes” for an EC key that can't do ECDSA? This is poorly defined at the moment: we don't explain what `mbedtls_pk_can_do` returns. I'm inclined to say that opaque engines may treat ECKEY, ECDH and ECDSA identically, since whatever they do there's no guarantee that the operation would work.